### PR TITLE
feat(inward): charge timed services per minute, per day or one time (#23206)

### LINKED ISSUE #23206: Time-Based Service – Need Per-Minute Calculation
File Path: /inward/inward_timed_item_fee.xhtml

**Description:**
For the added **Time Service**, the current functionality allows to calcualte in hoursly only. The service amount also needs to be calculated based on the actual duration in **minutes**. WHen the user created a timed service, he will have new input as how this is calculated, one time fee, hourly or by minutes or by days.

**Requirement:**

* Calculate the service charge according to the number of minutes used in addition to the existing hours-based one. Also need to add by days and one-time.
* The system should calculate the total amount automatically based on the configured **per-minute rate** when the selected service has that function. 
* The calculation should support different service durations should be availalbe under all places of time service calculation including the room management page and interim bill and final bill pages.

**Example of new minute based one:**

* Per-minute rate: Rs. 10
* Service duration: 30 minutes
* Total amount: **Rs. 300**

**Expected Result:**
The system should automatically calculate and display the correct service amount based on the actual number of minutes.

**Priority:** High

### PR DESCRIPTION
**Pull request by buddhika75**
Closes #23206

## Problem

A timed service fee could only be defined in hours. `TimedItemFee.durationHours`
was the length of one charged block and every calculation multiplied it by 60,
so:

- a per-minute service had to be faked as a fractional hour (`0.0166…`),
- a per-day service worked only because someone typed `24`,
- a flat one-off charge could not be expressed at all.

## Fix

`TimedItemFee` gains a `durationUnit` — `ONE_TIME | MINUTE | HOUR | DAY` — that
says what `durationHours` and `overShootHours` are counted in. The conversion
lives on the entity (`getDurationMinutes()`, `getOverShootMinutes()`,
`getDurationInHours()`), so the existing calculation funnel picks it up without
any call site changing its logic:

`InwardBeanController.calCount` / `calCountWithoutOverShoot` are what the room
management page, the interim bill, the final bill, the timed service consume
page and the surgery bill all go through. `ONE_TIME` short-circuits both to a
count of 1, so the fee is charged exactly once however long the service ran.

Configuration UI: **Charged By** dropdown next to the (renamed) **Duration**
field on `/inward/inward_timed_item_fee.xhtml`, plus the same selector on
`/inward/inward_room_facility.xhtml` for room charge blocks.

## Decisions made without approval

The issue gave the requirement but not the design. Each of these was decided
from the code rather than asked:

| Decision | Why |
|---|---|
| The unit goes on `TimedItemFee`, not `TimedItem` | `RoomFacilityCharge` holds a `TimedItemFee` directly with **no owning `TimedItem`**, so a service-level field could never reach room charges. The issue also names the *fee* page as the place to configure this. |
| `durationHours` keeps its name | Database compatibility. It now means "duration value"; the UI labels it "Duration" and pairs it with "Charged By". |
| `getDurationUnit()` returns `HOUR` when the column is null, instead of null | Every fee saved before this field existed was hour-based. Defaulting on read also stops the config dropdowns from rendering a legacy fee as whichever option is listed first — which would have silently converted rows to `ONE_TIME` on the next save. The entity hierarchy uses field access, so nothing is written back until a unit is actually chosen. |
| Saving a time-based fee with duration `0` is now rejected on the fee page | It could only ever price at zero, and `ONE_TIME` is the proper way to express a flat charge. |
| REST/AI surfaces extended too | Fee DTOs, the timed-item fee endpoints, the room facility charge endpoint and the capability statement — so the unit is not settable only through the UI. |
| A one-time room charge shows as a single billed slot in the room duration breakdown, and as its raw amount in the vacant-room "24 Hrs" column | Both are display mirrors of `calCount`, which returns 1 for a one-time fee. |

## ⚠️ Schema change required before deploy

`TimedItemFee` is part of the single-table `Fee` hierarchy, so this needs a new
**`DURATIONUNIT` column on the `FEE` table**. It has **not** been generated or
applied here — run `/generate-ddl` and apply it through the admin
**Add Missing Columns** page.

## Verification

- **14 new unit tests** (`TimedItemDurationUnitCalculationTest`), covering:
  - the issue's worked example — Rs. 10/minute × 30 minutes = **Rs. 300**
  - multi-minute blocks (15-min blocks, 50 minutes → 3 blocks)
  - per-day blocks, and an overshoot expressed in days (½ day grace on a 36-hour
    stay → 2 days)
  - one-time charging at 5 minutes, at a week, and at zero elapsed time
  - **regression guard**: a fee with no unit counts whole hour blocks exactly as
    before, and `HOUR` produces an identical result to the legacy null
- Full suite: **221 tests, 0 failures**.
- Browser verification was **not** possible in this unattended run: the
  `DURATIONUNIT` column does not exist in the local database yet, and applying
  DDL is deliberately left to a human. Confirmed the column is absent
  (`INFORMATION_SCHEMA` check on `FEE`) rather than assumed. Once the column is
  added, the paths worth clicking through are: configure a per-minute service →
  consume it on `/inward/inward_timed_service_consume.xhtml` → check the interim
  and final bill totals; and a room facility charge switched to `DAY`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Timed fees support one-time, minute, hour, and day-based billing.
  - Room and service fee forms allow selecting duration units.
  - APIs include duration units in requests and responses.
  - One-time fees can omit duration and overshoot values.

- **Bug Fixes**
  - Corrected calculations for varied duration units and one-time charges.
  - Added validation requiring positive durations for time-based fees.
  - Room charges now calculate daily equivalents consistently.
  - Charges use the entered service start time when available.
  - Updated fields refresh correctly when changing billing units.

- **Documentation**
  - Updated API, capability, and testing guidance.

### CODE REVIEW

**Comment by buddhika75** [comment]
## End-to-end verification done (local Payara + MySQL)

The `DURATIONUNIT` column was applied to the local `FEE` table and the branch was
deployed, so the earlier "not browser-verified" caveat no longer stands. Full write-up
with screenshots: **[Admin — Timed Service Charging Units](https://github.com/hmislk/hmis/wiki/Admin-Timed-Service-Charging-Units)**.

### What was exercised

Two timed services were created through the app and consumed against a real
undischarged admission, alongside a pre-existing hour-based service left untouched as a
regression control:

| Service | Charged By | Rate | Time used | Charged |
|---|---|---|---|---|
| Nebulisation | Per Minute, block 1 | 10 | 30 min | **300** ✅ the issue's worked example |
| Physiotherapy Visit | One Time Fee | 2,000 | 7 days | **2,000** ✅ |
| Nebulisation | Per Minute, block 1 | 10 | 46 min | **460** ✅ |
| Oxygen Therapy *(pre-existing, `DURATIONUNIT` still NULL)* | reads as Per Hour, block 1 | 500 | 2 h 30 min | **1,000** ✅ unchanged — 2 whole blocks |

Confirmed in the database (`patientitem.SERVICEVALUE`, `billitem.NETVALUE`) and on the
**interim bill → Timed Service Detail** tab, not just on the entry screen.

Also verified:
- Selecting **One Time Fee** disables Duration/Over Shoot via the AJAX listener.
- A `Per Hour` fee with Duration `0` is rejected and the message is now visible.
- **Room charges**: a vacant room switched from `Per Hour / 24` to `Per Day / 1`
  produces an identical daily equivalent (3,000.00 either way); it was restored
  afterwards.

### Two defects the testing exposed — both fixed in d5264f9

1. **`save()` priced from the date of admission, not the entered Start Time.**
   Pre-existing since Feb 2024 and inconsistent with the row's own Update button, which
   prices from `fromTime`. Per-minute charging made it unmissable: the first save of a
   6-minute service on a 2-week-old admission billed **Rs. 217,720** instead of Rs. 60.
   Now prices from `fromTime`, falling back to the admission date only when no start
   time was recorded.

2. **The fee page swallowed its own FacesMessages.** The Add button updated only
   `gpDetail dtbFee`, so nothing — including the new duration validation — ever reached
   the growl. Added `:growl` to the update.

### Not fixed (pre-existing, out of scope — worth a separate issue)

The interim bill's **Timed Service Detail** table has a second `Service Name` column
rendering the entity's `toString()` (`com.divudi.core.entity.PatientItem[ id=7 ]`)
instead of a property. Unrelated to charging units; left alone deliberately.

### Still required before deploy

The `DURATIONUNIT` column on `FEE` — run `/generate-ddl` and apply through the admin
**Add Missing Columns** page. It was applied to the local database only.

CI: all checks passing on the previous commit; re-running for d5264f9.

File: src/main/java/com/divudi/bean/inward/InwardBeanController.java [hunk 1]
```diff
@@ -3323,7 +3323,12 @@ public double calTotalLinen(PatientEncounter patientEncounter) {
 
     public double calCountWithoutOverShoot(TimedItemFee tif, Date admittedAt, Date dischargedAt) {
 
-        double duration = tif.getDurationHours() * 60;
+        // A one-time fee is charged once for the whole service, however long it ran.
+        if (tif != null && tif.isOneTime()) {
+            return 1;
+        }
+
+        double duration = tif.getDurationMinutes();
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/bean/inward/InwardBeanController.java:3339]: Valid — fixed in 3541a96.

You are right that the partial `tif != null &&` advertised null-safety the method did not have. Both `calCount` and `calCountWithoutOverShoot` now guard the whole method.

They return **0** rather than throwing, which is a deliberate choice worth flagging: `RoomFacilityCharge.timedItemFee` is a nullable `@ManyToOne` with no DB constraint, so a null is reachable from the room paths, and an NPE there breaks the page that is rendering the bill. Returning 0 also matches what the method already does for a fee whose duration is 0 — an unconfigured fee and a missing fee now behave the same way. No such rows exist in the database I tested against (65 room charges, 0 with a null fee), so this is defensive rather than a live bug.

Covered by a new unit test (`aMissingFeeCountsAsNothingToBillRatherThanThrowing`).

File: src/main/webapp/inward/inward_timed_item_fee.xhtml [hunk 2]
```diff
@@ -146,11 +167,19 @@
...
+                                                           itemValue="#{du}"/>
+                                        </p:selectOneMenu>
+                                    </p:column>
+                                    <p:column headerText="Duration">
+                                        <p:inputText value="#{bi.durationHours}" autocomplete="off" disabled="#{bi.oneTime}" />
                                     </p:column>
                                     <p:column headerText="Over Shoot">
-                                        <p:inputText value="#{bi.overShootHours}" autocomplete="off" />
+                                        <p:inputText value="#{bi.overShootHours}" autocomplete="off" disabled="#{bi.oneTime}" />
                                     </p:column>
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/webapp/inward/inward_timed_item_fee.xhtml:187]: Valid, and the more serious of the two — fixed in 3541a96.

Confirmed the trap exists exactly as described. Applied your fix (change listener on the row dropdown, ids on the two inputs it refreshes) and added a server-side backstop: `updateFee()` now carries the same duration guard `saveCharge()` already had. The client-side enable/disable is a convenience; the server check is what actually prevents a time-based fee with a duration of 0, whichever way the row was edited.

Verified on a running deployment against a real One Time fee row:

1. Row renders with Duration/Over Shoot disabled — `rowDuration`/`rowOverShoot` ids resolve correctly inside the datatable row.
2. Switching the row to Per Hour re-enables both inputs immediately.
3. Pressing Update with the duration still 0 is rejected; the stored fee stayed `ONE_TIME` / 0 rather than becoming a zero-priced hourly fee.
4. Entering a real duration and pressing Update saves `HOUR` / 2 normally.

Also added a unit test for a one-time fee carrying a leftover non-zero duration, since switching a row back to One Time is now a state the page can produce.

**Comment by buddhika75** [comment]
## Review cycle 2 — both outside-diff findings addressed (6e52bf7)

**`calCountWithoutOverShoot` divided by a zero duration.** Valid. `calCount`
already guarded on `duration != 0`; this one didn't, so a time-based fee with no
duration yielded `Infinity`, which casts to a huge block count — an *overcharge*,
not a crash. Now returns 0 to match. Worth noting for scale: the method has **no
callers** anywhere in the codebase, so this is hardening, not a live defect.

**Room facility charge endpoint accepted a block that can never bill.** Valid,
and the more consequential of the two. Added the same rule the sibling
`/timed-items/{id}/fees` endpoint already enforces.

⚠️ **This tightens an existing API contract — please sanity-check this call.**
`timedItemFeeDurationHours` was documented optional, and omitting it produced a
fee with `durationHours = 0`. My reasoning for tightening rather than leaving it:
such a charge was *already* useless — the request returned 200 but the room then
billed nothing for every stay, forever, silently. Rejecting it surfaces an
already-broken call rather than breaking a working one, and `ONE_TIME` remains
the supported way to express a flat charge with no duration. `API_INWARD_ROOM.md`
updated to match. If you'd rather keep the endpoint permissive, this is the one
commit to revert.

### Verified against the running endpoint

| Request | Result |
|---|---|
| `POST` unit `HOUR`, duration `0` | **400** — "must be greater than 0 for a HOUR block" |
| `POST` duration omitted (defaults to `HOUR`) | **400** |
| `POST` unit `ONE_TIME`, no duration | **200** |
| `POST` unit `DAY`, duration `1` | **200** |
| `PUT` switching `ONE_TIME` → `HOUR` with no duration | **400** |
| `PUT` switching `ONE_TIME` → `HOUR` with duration `12` | **200** |

Probe records created for those checks were deleted afterwards. Unit tests now
17 in `TimedItemDurationUnitCalculationTest`, **224 total, all passing**.

File: src/main/java/com/divudi/ws/inward/InwardRoomFacilityChargeApi.java [hunk 3]
```diff
@@ -250,6 +252,12 @@ public Response create(String requestBody) {
             if (overShootHours != null) timedItemFee.setOverShootHours(overShootHours);
             Long durationDaysForMoCharge = asLong(body.get("timedItemFeeDurationDaysForMoCharge"));
             if (durationDaysForMoCharge != null) timedItemFee.setDurationDaysForMoCharge(durationDaysForMoCharge);
+            TimedItemDurationUnit durationUnit = asDurationUnit(body.get("timedItemFeeDurationUnit"));
+            if (durationUnit != null) timedItemFee.setDurationUnit(durationUnit);
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/ws/inward/InwardRoomFacilityChargeApi.java:260]: Applied in f5f5c95 — for both API surfaces, not just this one.

One correction to the framing though: there is **no behavioural difference**. Every read goes through `getDurationUnit()`, which resolves null to `HOUR`, so an API-created fee with a null column and a UI-created fee with `'HOUR'` price identically and both report `"durationUnit": "HOUR"` in the response. This is a data-consistency tidy-up, not a defect.

It is still worth doing: the fee page and the room charge page each write `HOUR` explicitly for a new record, so identical fees were being stored two different ways depending on where they were created. New rows from either API now carry the explicit value too.

Worth being clear about what this does *not* achieve: rows created before duration units existed stay null permanently — deliberately, since that null is the backward-compatibility default the whole design rests on. Anything reading `DURATIONUNIT` directly still has to treat null as HOUR; this just stops adding new ambiguity.

Verified against the running endpoints: a room facility charge POSTed without `timedItemFeeDurationUnit`, and a timed-item fee POSTed without `durationUnit`, both now store `HOUR`. Probe records deleted afterwards.
